### PR TITLE
Benchmarks extension

### PR DESCRIPTION
**Pull request by jsmassa**
New queries and more options to restrict evaluation

### CODE REVIEW

**Comment by kordano** [comment]
Could you add the docs for running the benchmarks in here as well? That would close #327.

**Comment by jsmassa** [comment]
I would add the doc as soon as I am sure, they are not changing for a while. Since the optimizations are not done yet, this is not the case yet though

**Comment by jsmassa** [comment]
But I think there is also no a rush to merge this branch, so we could also wait until the docs are in. I think that should be done within the next month.

File: deps.edn [hunk 1]
```diff
@@ -38,7 +38,9 @@
            :benchmark {:main-opts ["-m" "benchmark.core"]
                        :extra-paths ["benchmark/src"]
                        :extra-deps {clj-http/clj-http     {:mvn/version "3.11.0"}
-                                    org.clojure/tools.cli {:mvn/version "1.0.194"}}}
+                                    org.clojure/tools.cli {:mvn/version "1.0.194"}
+                                    incanter/incanter-core {:mvn/version "1.9.2"}
```
Comment (kordano) [inline deps.edn:42]: incanter should also be added to the `bench-dev` environment.

File: benchmark/src/benchmark/core.clj [hunk 2]
```diff
@@ -9,38 +9,90 @@
...
+   ["-f" "--function FUNCTIONNAME"  
+    (str "Name of function to test. Available are " implemented-functions)
+    :default :all
+    :parse-fn read-string
+    :validate [implemented-functions (str "Must be a keyword for implemented database function. "
+                                          "Available are: " implemented-functions)]]
+
+   ;; CMD compare
+
+   ["-p" "--[no-]plots" "Output comparison as plots."]
```
[the remark this answers is not in the document]
Comment (kordano) [reply benchmark/src/benchmark/core.clj:91]: And what about plots for the normal benchmarks?

**Review by kordano — CHANGES_REQUESTED** [review]
There should be a test namespace to check whether the benchmarks are working correctly

File: benchmark/src/benchmark/core.clj (hunk 2 above)
[the remark this answers is not in the document]
Comment (jsmassa) [reply benchmark/src/benchmark/core.clj:91]: Is there a point to having them without comparing something in the plot?

For the comparisons you would just run them for one branch, usually the branch you forked from, then run them in a different branch, that would be your feature branch, give them different tags of course, and then run benchmarks compare filename1 filename2 -p for plots or without the -p for tables.

File: benchmark/src/benchmark/config.clj [hunk 3]
```diff
@@ -23,51 +23,63 @@
...
-             :keep-history? true
-             :index :datahike.index/persistent-set
-             :name "mem-set"}}
+             :keep-history? false
+             :index :datahike.index/persistent-set}}
    {:config-name "mem-hht"
     :config {:store {:backend :mem :id "performance-hht"}
              :schema-flexibility :write
-             :keep-history? true
+             :keep-history? false
```
[the remark this answers is not in the document]
Comment (jsmassa) [reply benchmark/src/benchmark/config.clj:31]: There is nothing history-specific being tested, so my thinking was that without it, it is simpler. But I am not fussed about what it is, so if you think we should benchmark it with history, we can do that

File: benchmark/src/benchmark/measure.clj [hunk 4]
```diff
@@ -4,72 +4,91 @@
...
-             {:time t-transaction-n :context {:dh-config simple-config :function :transaction :db-entities initial-entities :db-datoms initial-datoms 
-                                              :execution {:tx-entities tx-entities :tx-datoms tx-datoms}}}
-             {:time t-connection-0n  :context {:dh-config simple-config :function :connection  :db-entities final-entities :db-datoms final-datoms}}])))
+        _ (d/delete-database unique-config)
+        transaction-times (if (#{:all :transaction} function)
+                            (for [tx-entities tx-entity-counts
+                                  _ (range iterations)]
+                              (let [unique-config (assoc config :name (str (UUID/randomUUID)))
+                                    {:keys [conn]} (init-db db-entities unique-config)
+                                    tx (vec (repeatedly tx-entities c/rand-entity))
```
[the remark this answers is not in the document]
Comment (jsmassa) [reply benchmark/src/benchmark/measure.clj:79]: I see, guess I didn't run it for transaction for transactions in a long time

**Comment by jsmassa** [comment]
Closes #327 now

File: deps.edn (hunk 1 above)
Comment (jsmassa) [reply deps.edn:48]: Now added to dev

File: benchmark/src/benchmark/config.clj [hunk 5]
```diff
@@ -23,51 +23,63 @@
...
+    :db/valueType   :db.type/string
+    :db/cardinality :db.cardinality/one}
+   {:db/ident       :i1
+    :db/valueType   :db.type/bigint
+    :db/cardinality :db.cardinality/one}
+   {:db/ident       :i2
+    :db/valueType   :db.type/bigint
+    :db/cardinality :db.cardinality/one}])
+
+(defn rand-str [max-int]
```
[the remark this answers is not in the document]
Comment (jsmassa) [reply benchmark/src/benchmark/config.clj:51]: Removed the constant now

File: deps.edn [hunk 6]
```diff
@@ -1,6 +1,6 @@
 {:deps {org.clojure/clojure                         {:mvn/version "1.10.1"}
         org.clojure/clojurescript                   {:mvn/version "1.10.597"}
-        io.replikativ/hitchhiker-tree               {:mvn/version "0.1.11"}
+        io.replikativ/hitchhiker-tree               {:mvn/version "0.1.10"}
```
Comment (kordano) [inline deps.edn:3]: what is the reason to use the former hitchhiker-tree?

File: deps.edn [hunk 7]
```diff
@@ -52,11 +56,6 @@
            :build {:extra-deps {seancorfield/depstar {:mvn/version "1.1.136"}}
                    :main-opts ["-m" "hf.depstar.jar" "replikativ-datahike.jar"]}
 
-           :install {:extra-deps {slipset/deps-deploy {:mvn/version "0.1.5"}}
```
Comment (kordano) [inline deps.edn:55]: Why are you removing install?

File: benchmark/test/benchmark/measure.clj [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+             :iterations 1, 
+             :data-found-opts :all
+             :tx-entity-counts [0 10]
+             :tag #{test} 
+             :query :all 
+             :db-entity-counts [0 10]})
+
+(deftest connection-test 
+  (let [measurements (b/get-measurements (assoc config :function :connection))]
+    (is (= (set (keys (:time (first measurements))))
```
Comment (kordano) [inline benchmark/test/benchmark/measure.clj:17]: For the future tests: the value you are expecting is the first argument. Some testing tools prefer this order in order to tell expectation vs actual apart.

File: doc/benchmarking.md [hunk 9]
```diff
@@ -0,0 +1,319 @@
...
+
+#### Queries (-q)
+
+Options for `-q`: 
+
+- Simple query: `:simple-query`
+- Join queries: `:e-join-query` `:e-join-query-first-fixed` `:e-join-query-second-fixed` `:a-join-query` `:v-join-query`
+- Predicate queries: `:equals-query` `:equals-query-1-fixed` `:less-than-query` `:less-than-query-1-fixed`
+- Queries using arguments from bindings: `:scalar-arg-query` `:scalar-arg-query-with-join` `:vector-arg-query`
+- Aggregater queries: `:stddev-query` `:variance-query` `:max-query` `:median-query` `:avg-query`
```
Comment (kordano) [inline doc/benchmarking.md:94]: Aggregator or Aggregate?

**Review by kordano — CHANGES_REQUESTED** [review]

**Comment by jsmassa** [comment]
All requested changes fixed now

**Review by kordano — APPROVED** [review]
Looks good, thanks.

**Pull request merged**